### PR TITLE
Parse introspection types separately from the provided schema

### PR DESCRIPTION
**Pull request by SimonSapin**
Fixes https://github.com/apollographql/router/issues/2269

**Checklist**

Complete the checklist (and note appropriate exceptions) before a final PR is raised.

- [x] Changes are compatible[^1]
- [x] Documentation[^2] completed
- [x] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]. It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]. Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]. Tick whichever testing boxes are applicable. If you are adding Manual Tests:
    - please document the manual testing (extensively) in the Exceptions.
    - please raise a separate issue to automate the test and label it (or ask for it to be labeled) as `manual test`

### CODE REVIEW

File: apollo-router/src/error.rs [hunk 1]
```diff
@@ -423,7 +423,7 @@ struct ParserError {
 impl ParseErrors {
     #[allow(clippy::needless_return)]
     pub(crate) fn print(&self) {
-        if LevelFilter::current() == LevelFilter::OFF {
+        if LevelFilter::current() == LevelFilter::OFF && cfg!(not(debug_assertions)) {
             return;
```
Comment (SimonSapin) [inline apollo-router/src/error.rs:427]: This hack is to make tests take the potentially-panicky path

**Review by bnjjj — APPROVED** [review]

**Comment by SimonSapin** [comment]
I’d missed that `document.definitions()` was used in multiple places, please take a look at the fix. (New commit to be squashed)

**Review by bnjjj — APPROVED** [review]

File: apollo-router/src/error.rs (hunk 1 above)
Comment (Geal) [reply apollo-router/src/error.rs:427]: could you explain a bit more this change and the next on line 436? The rest of the PR looks nice but that part feels weird

Comment (SimonSapin) [reply apollo-router/src/error.rs:427]: There was a panicky bug. I want first to reproduce the panic in a test, then fix it. There were two obstacles to reproduction:

* `LevelFilter::current()` appears to be OFF while running test, so the panicky code path was not taken at all. This change causes some logging to stdout to happen even if logging is nominally disabled, when compiled with debug assertions such as for testing. In the test harness however, stdout is captured by default and only shown when a test fails. So this doesn’t normally add visible output.

  I don’t really like this weird condition, but it’ll goo away when we drop AST-based parsing. This shouldn’t print to stdout in the first place. This PR fixes the panic (and tests the fix) without waiting for that larger refactor.

* The second obstacle is https://github.com/rust-lang/rust/issues/107118 which would swallow the panic when stdout is captured by the test harness. As commented, adding `format!` works around that.

**Pull request merged**